### ISSUE TITLE
Security-Issue sonatype-2018-0590 (possible XXE (eXternal XML Entity)-attack when Using XML as Data Source)

### ISSUE THREAD

**Issue by mr0reed137** [eclipse-birt/birt#1986]
Hello,

Our security-scan recently reported a security vulnerability in the rhino library, namely **sonatype-2018-0590** with the following explanation:

<img width="354" alt="2024-11-26 10_29_56-IQ Server - Component Details – Mozilla Firefox" src="https://github.com/user-attachments/assets/c047e81f-885f-4383-9918-0d525a6e0179">

This library is included in the Birt Engine (v4.14.0). 

I can confirm that the Birt Engine processes XMLs extended by XXE accordingly and that the vulnerability can therefore be exploited.

An example for a manipulated XML with XXE-attack:

<img width="156" alt="2024-11-26 10_41_02-New Issue · eclipse-birt_birt – Mozilla Firefox" src="https://github.com/user-attachments/assets/3dbd9e15-f81b-4024-85fd-40982f79d936">

The result of the exploited vulnerability in this case is that the contents of the /etc/passwd - file are printed in the pdf.

Although the rhino library has been updated in Birt version 4.17, the vulnerability can still be exploited. I suspect that the processing of XMLs in the Birt Engine as a datasource is not implemented with rhino.

Is this known and can the processing of XML with external XML entities be prevented?

**Comment by merks** [comment]
I'm not sure the basis for asserting that the exploit you describe is in some way related to the version of rhino that's being used.  The issue that you show says 1.7.12 fixes the problem and given BIRT 4.17 uses 1.7.15  there is some kind of contradiction or confusion there.

As @hvbtup suggests, if the sources of your report inputs are not secure then you need to secure those inputs such that they do not contain content that can comprise your server or can produce compromised reports.

I expect there are places where BIRT processes XML where we could and probably should disable some of the vulnerable aspects in the existing XML libraries, e.g., entity resolution and external DTDs.  But who is going to do that work or fund that work?  I expect no one will do either.

**Comment by hvbtup** [comment]
As @merks said, based on the information you gave us, the issue is already fixed in 4.17.

**Comment by mr0reed137** [comment]
The issue is fixed in rhino, yes. 
But the XML-processing, as @merks mentioned, is obviously not done with rhino at all. And that is what I can confirm, because when using XML as datasource in Birt reports, the security issue can still be exploited - even in Birt 4.17. So there is another issue in some java library, but as I understand one would have to analyze how XML is processed and fix it there (but this is unlikely to happen due to missing resources/capacities).

**Comment by merks** [comment]
I don't think the are other places directly in BIRT but in upstream dependencies there likely are other similar problems.  But they will need to be addressed upstream; investment in free goodness is required.

**Comment by mr0reed137** [comment]
Thank you...will the fix be available in Birt 4.18 then?

**Comment by merks** [comment]
Yes, this particular change will be available, in the release and is already available here:

https://download.eclipse.org/birt/updates/nightly/latest/

Note that I don't assure that this change addresses the actual problem you describe which is more likely to come from the datatools project:

https://github.com/eclipse-datatools/datatools/issues/22

**Comment by mr0reed137** [comment]
I will try to check the nightly build, if I can reproduce the problem or if it is fixed and give feedback here.

**Comment by hvbtup** [comment]
Ed's fix seems to work (see my comment in https://github.com/eclipse-datatools/datatools/issues/22). The error handling could be improved, but that can be done later.

**Comment by mr0reed137** [comment]
As suggested I downloaded the nightly build (which was from yesterday --> birt-runtime-osgi-4.18.0-202411271252) and I still can reproduce the problem. Was the fix included in that build?

**Comment by merks** [comment]
This build **from today** has also the datatools changes:

https://download.eclipse.org/birt/updates/nightly/N202411281200

Please try that one.
